### DISCUSSION TITLE
How to manually add files with 2.0

### DISCUSSION THREAD

**Discussion by JustSimplyKyle** [fzyzcjy/flutter_rust_bridge#1450] [Q&A] [answered]
I've found this file.flutter_rust_bridge.yaml
```yaml
rust_input: rust/src/api{.rs,/authentication/*.rs}
dart_output: lib/src/rust
```

but it can't find it...
```bash
Error: Find zero rust input paths. (glob_pattern="/home/kyle/coding/era-connect/app/rust/src/api{.rs,/authentication/*.rs}")
error: Recipe `gen` failed on line 5 with exit code 1
```

**Comment by fzyzcjy** [comment]
Hmm, the default one supports *everything* inside `api` folder (should be recursively as well, if not it is a bug). Does that work for you?

**Reply by JustSimplyKyle** [reply]
yea, that works, but I don't wanna add all my stuff, since it wouldn't compile(compains about not supporting `AsRef`)

**Reply by fzyzcjy** [reply, marked as answer]
Ah, I see. 

1. IMHO it is a common practice to have "layered architecture", e.g. one layer (folder) for the api that FRB will convert, another layer for internal implementation, etc.
2. Could you please make a minimal reproducible sample about `AsRef`? Maybe it is something that the `arbitrary rust types` can support.
3. Also feel free to enhance the `rust_input`! (IIRC it is a simple `glob` currently; but we may discuss what to add before implementing it, e.g. a regex may be too powerful?)

**Reply by JustSimplyKyle** [reply]
1. Thanks! Currently it's not layered, all folders are one layer deep in api, will try to refactor!
2. Will do later, should be quite easy to do a reproducible example
3. I'll take a stab at it once I migrated it to 2.0

**Reply by JustSimplyKyle** [reply]
here
```rust
#[flutter_rust_bridge::frb(sync)] // Synchronous mode for simplicity of the demo
pub fn greet(name: impl AsRef<str>) -> String {
    format!("Hello, {}!", name.as_ref())
}
```
bash
```
❯ flutter_rust_bridge_codegen generate        
[0.1s] Parse  
  └── [0.1s] Run cargo expand  
  └── [0.0s] Parse source graph                                                                                               Error: function=Ident { sym: greet, span: bytes(37..42) }

Caused by:
    0: when trying to parse DartFn
    1: Unknown ident: AsRef
```

**Reply by fzyzcjy** [reply]
Ah I see, maybe because `impl Asref<str>` confuses it. Currently `impl` has only one scenario - `impl Fn(A) -> B` will be converted such that, a Dart function can be passed in.

Anyway, here, the case is quite special: We indeed want FRB to treat it as a `String`, and not a rust-auto-opaque type.
So for this one special case, maybe we can let FRB understand that, when seeing `impl asref<str>`, it should think as a String. But I am not sure whether this is too special for this one case and not general enough.

**Reply by JustSimplyKyle** [reply]
I think `AsRef<T>` is used for `Path` `str` and other similar non owned types
so maybe we can get the owned type via `.as_ref().to_owned()`(altho I'm not sure whether if this method can work in a code generating environment), and use that as the type

**Reply by fzyzcjy** [reply]
I agree. Maybe we should generate the owned type for the Dart side, but via hardcoding (since I guess `.as_ref.to_owned` may not work in codegen).
